.circleci/config.yml

```diff
@@ -1,115 +1,163 @@
 version: 2.1
 orbs:
-  android: circleci/android@1.0
+  android: circleci/android@2.1.2
+  codecov: codecov/codecov@3.2.4
 
 # Workflows orchestrate a set of jobs to be run;
 workflows:
   version: 2
   build-test-deploy:
     jobs:
-      - validate-code
+      - validate-code:
+          gradle-cache-prefix: v1
+          build-cache-prefix: v1
+
       - build-and-unit-test:
+          gradle-cache-prefix: v1
+          build-cache-prefix: v1
           requires:
             - validate-code
+      
       - functional-test:
+          gradle-cache-prefix: v1
+          build-cache-prefix: v1
           requires:
             - validate-code
- 
+
 jobs:
   validate-code:
-    working_directory: ~/code
-    docker:
-      - image: circleci/android:api-29-node
-    environment:
-      JVM_OPTS: -Xmx3200m
+    parameters:
+      gradle-cache-prefix:
+        type: string
+        default: v1
+      build-cache-prefix:
+        type: string
+        default: v1
+
+    # List of available Android Docker images: https://circleci.com/developer/images/image/cimg/android#image-tags
+    executor:
+      name: android/android-machine
+      resource-class: large
+      tag: 2022.09.1
```

Comment:
We don't need a large Android machine image here as we are not using an emulator. The Android Docker image is fine. Can you change to 
```
    executor:
      name: android/android-docker
      tag: 2024.01.1
```
If that version of the image doesn't support some of our current libs, such as Spotless, then use `2022.09.2-node` for now and we can upgrade later when the Gradle files are updated.